repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -106,20 +100,34 @@ class Session(interpreter: Interpreter)
     _statements.clear()
   }
 
+  private def changeState(newState: SessionState): Unit = {
+    synchronized {
+      _state = newState
+    }
+    stateChangedCallback(newState)
+  }
+
   private def executeCode(executionCount: Int, code: String): String = synchronized {
-    _state = SessionState.Busy()
+    changeState(SessionState.Busy())
+
+    def transitToIdle() = {
+      val executingLastStatement = executionCount == newStatementId.intValue() - 1
+      if (_statements.isEmpty || executingLastStatement) {
```

Comment:
I don't disagree with you. But I'm hesitate to remove this now. I think we should keep the busy state. Start a conversation in livy-user@ mailing list and ask our users what's their takes on this.